crates/core/tedge_actors/src/builders.rs

```diff
@@ -0,0 +1,232 @@
+use crate::mpsc;
+use crate::ClientId;
+use crate::ConcurrentServiceMessageBox;
+use crate::DynSender;
+use crate::KeyedSender;
+use crate::Message;
+use crate::NullSender;
+use crate::RequestResponseHandler;
+use crate::RuntimeError;
+use crate::RuntimeHandle;
+use crate::Sender;
+use crate::SenderVec;
+use crate::ServiceMessageBox;
+use crate::SimpleMessageBox;
+use async_trait::async_trait;
+use std::convert::Infallible;
+
+/// Materialize an actor instance under construction
+///
+/// Such an instance is:
+/// 1. built from some actor configuration
+/// 2. connected to other peers
+/// 3. eventually spawned into an actor.
+#[async_trait]
+pub trait ActorBuilder {
+    /// Build and spawn the actor
+    async fn spawn(self, runtime: &mut RuntimeHandle) -> Result<(), RuntimeError>;
+}
+
+/// Builder of `T`
+pub trait Builder<T>: Sized {
+    type Error: std::error::Error;
+
+    /// Builds the entity or returns an error
+    fn try_build(self) -> Result<T, Self::Error>;
+
+    /// Builds the entity or panics
+    fn build(self) -> T {
+        self.try_build().unwrap()
+    }
+}
+
+/// A trait to connect a message box under-construction to peer messages boxes
+pub trait MessageBoxConnector<Request: Message, Response: Message, Config: Default> {
```

Comment:
Wouldn't that `Config: Default` constraint force every `Config` type to have a default? That wouldn't be feasible all the time, right? Wouldn't it be better to provide a `NullConfig` type as part of the API rather than this constraint?

Reply:
What do you mean by a `NullConfig` type?

Reply:
Okay, I see. This is indeed necessary for any kind of connection that requires no config. But this is not enough for what I have in mind: removing the burden of passing a unit `()` when  creating such connections. Adding a `Ext` implemented only when there is default config can be an option.

```
pub trait MessageBoxConnector<Request: Message, Response: Message, Config> {
    fn connect_with(&mut self, peer: &mut impl MessageBoxPort<Request, Response>, config: Config);
}

impl<T: ?Sized, I,O,C> MessageBoxConnectorExt<I,O> for T where T: MessageBoxConnector<I,O,C> {
    /// Use the default config to connect a peer message box
    fn connect(&mut self, peer: &mut impl MessageBoxPort<Request, Response>) {
        self.connect_with(peer, Config::default())
    }
}

put trait MessageBoxConnectorExt<Request, Response> {
    fn connect(&mut self, peer: &mut impl MessageBoxPort<Request, Response>);
}
```

However, this starts to be a lot of traits, while having a default for a config seems reasonable.

Reply:
I removed the `Default` constraint on `Config`.